compiler/rustc_trait_selection/src/traits/util.rs

```diff
@@ -506,20 +507,80 @@ impl<'tcx> TypeFolder<TyCtxt<'tcx>> for PlaceholderReplacer<'_, 'tcx> {
         predicate.kind().skip_binder()
     {
-        if tcx.is_lang_item(trait_ref.def_id(), LangItem::Sized)
-            && trait_ref.self_ty().is_trivially_sized(tcx)
+        let sizedness = if tcx.is_lang_item(trait_ref.def_id(), LangItem::Sized) {
+            Some(SizedTraitKind::Sized)
+        } else if tcx.is_lang_item(trait_ref.def_id(), LangItem::MetaSized) {
+            Some(SizedTraitKind::MetaSized)
+        } else if tcx.is_lang_item(trait_ref.def_id(), LangItem::PointeeSized) {
+            Some(SizedTraitKind::PointeeSized)
+        } else {
+            None
+        };
+
+        if let Some(sizedness) = sizedness
+            && trait_ref.self_ty().has_trivial_sizedness(tcx, sizedness)
         {
             debug!("fast path -- trivial sizedness");
             return true;
         }
+
+        if matches!(sizedness, Some(SizedTraitKind::MetaSized)) {
+            let has_sized_in_param_env =
+                param_env.caller_bounds().iter().filter_map(|c| c.as_trait_clause()).any(|c| {
+                    trait_ref.self_ty() == c.skip_binder().self_ty()
+                        && tcx.is_lang_item(c.def_id(), LangItem::Sized)
+                });
+            if has_sized_in_param_env {
+                debug!("fast path -- metasized paramenv");
+                return true;
+            }
+        }
     }
 
     false
 }
+
+/// To improve performance, sizedness traits are not elaborated and so special-casing is required
+/// in the trait solver to find a `Sized` candidate for a `MetaSized` predicate. This is a helper
+/// function for that special-casing, intended to be used when the obligation to be proven has been
+/// confirmed to be `MetaSized`, and checking whether the `param_env` contains `Sized` for the same
+/// `self_ty` is all that remains to be done.
+pub(crate) fn is_unelaborated_sizedness_optimisation<'tcx>(
+    infcx: &InferCtxt<'tcx>,
+    predicate: PolyTraitPredicate<'tcx>,
+    candidates: impl IntoIterator<Item = PolyTraitPredicate<'tcx>>,
+) -> bool {
+    find_unelaborated_sizedness_optimisation(infcx, predicate, candidates).is_some()
+}
+
+/// See `is_unelaborated_sizedness_optimisation`, but returns the `PolyTraitPredicate` that worked.
+pub(crate) fn find_unelaborated_sizedness_optimisation<'tcx>(
+    infcx: &InferCtxt<'tcx>,
+    predicate: PolyTraitPredicate<'tcx>,
+    candidates: impl IntoIterator<Item = PolyTraitPredicate<'tcx>>,
+) -> Option<PolyTraitPredicate<'tcx>> {
+    if !infcx.tcx.is_lang_item(predicate.def_id(), LangItem::MetaSized) {
+        return None;
+    }
+
+    candidates.into_iter().find(|c| {
+        if !infcx.tcx.is_lang_item(c.def_id(), LangItem::Sized) {
+            return false;
+        }
+
+        let expected_self_ty = infcx.resolve_vars_if_possible(predicate.self_ty());
+        let found_self_ty = infcx.resolve_vars_if_possible(c.self_ty());
+        predicate.polarity() == c.polarity() && expected_self_ty == found_self_ty
```

Comment:
again, structurally comparing types is insufficient if there are aliases/regions

This fast path means we fail to prove `MyType<'_, T>: MetaSized` if there's `MyType<'a, T>: Sized` in the env

Reply:
Fixed and added test